parking/backend/engine/alloc_engine.py

```diff
@@ -0,0 +1,54 @@
+import logging
+from typing import Optional
+from geopy.distance import distance
+from parking.backend.db.dbaccess import DbAccess
+from parking.shared.rest_models import ParkingLot
+from parking.shared.location import Location
+from parking.shared.ws_models import ParkingRequestMessage
+
+logger = logging.getLogger('backend')
+
+
+class AllocationEngine():
+    def __init__(self, dba: DbAccess, user_sessions: 'UserSessions') -> None:
+        self.dba = dba
+        self.user_sessions = user_sessions
+
+    async def handle_allocation_request(self, user_id: str, request: ParkingRequestMessage) -> Optional[ParkingLot]:
+        rejections = self.user_sessions.get_user(user_id).rejections
+        lots = await self.dba.get_available_parking_lots(request.location, 100, rejections)
+
+        # Just take the first lot for now...
+        if lots:
+            lot = lots[0]
+            return ParkingLot(lot['capacity'], lot['name'], lot['price'], Location(lot['lat'], lot['long']))
+        else:
+            return lots[0] if lots else None
+
+    async def commit_allocation(self, user_id: str, park_id: int) -> bool:
+        success = await self.dba.allocate_parking_lot(user_id, park_id)
+        if not success:
+            logger.warning('Failed to allocate parking lot {} to user {}'.format(park_id, user_id))
+        return success
+
+    async def remove_allocation(self, user_id: str) -> None:
```

Comment:
I also need a 'delete_allocation(user_id)' function to call from the Websocket handler.

Reply:
Done :+1: